### ISSUE TITLE
RNN Stateful Backpropagation

### ISSUE THREAD

**Issue by HenryJia** [fchollet/keras#3669]
Suppose I loop the fit call manually on a stateful network with batch_input_shape = (1, 1, N) 

Will backpropagation still backprop through all the past states to get the full gradient? Or will it just treat the state as another input which it can't backprop through?

**Comment by carlthome** [comment]
Backprop will be truncated to 1 timestep with `batch_input_shape=(1, 1, N)` as the second dimension is 1. No gradient updates will be performed further back in time than the second dimension's value.

**Comment by marynelv** [comment]
If training is done with model.train_on_batch, Is backprop still truncated to 1 timestep with batch_input_shape=(B, 1, N) ?

**Comment by carlthome** [comment]
Yes.

**Comment by ParthaEth** [comment]
Wouldn't it be better if the error can be back propagated to the time stamp when the network was last reset? Since inputs from that time will clearly influence the output? Also this will make programming RNNs of the following kind way easier. Given this all the tutorials out there in the internet are actually wrong! There should be official examples for target sequences. Further more if the training sequences are of not equal length keras need to fall back on hacks. That's not really nice. Try to imagine how you would train the following net and you will see what I mean.
![simplernn](https://cloud.githubusercontent.com/assets/10944728/20348882/16b3f224-ac08-11e6-9cad-d33432405e9c.jpg)

**Comment by farizrahman4u** [comment]
What is special about it? (Assuming the xt-1 on top is a typo)

**Comment by carlthome** [comment]
@ParthaEth, it's an implementation detail. Computers cannot fit weight matrices of arbitrary sizes, and we prefer to do computations with matrices for computational efficiency (instead of something like a linked list). You're of course free to implement a RNN that can backprop correctly back to a timestamp, but it won't be fast.
